Remove the boost/algorithm/string/case_conv.hpp dependency

l2a5b1 opened the pull request:
This pull request removes the `boost/algorithm/string/case_conv.hpp` dependency from the project.

`boost/algorithm/string/case_conv.hpp` is included for the `boost::to_lower` and `boost::to_upper` template functions.

We can replace the calls to these functions with straightforward alternative implementations that use the C++ Standard Library, because the functions are called with `std::string` objects that use standard 7-bit ASCII characters as argument.

The refactored implementation should work without the explicit `static_cast<unsigned char>` cast and `unsigned char` lambda return type. Both have been added defensively and to be explicit. Especially in case of the former, behaviour is undefined (potentially result in a crash) if the `std::toupper` argument is not an `unsigned char`.

A potential alternative, maybe even preferred, implementation to address the `boost::to_lower` function call in `ParseNetwork(std::string)` could have been:

```c++
if (net == "ipv4" || net == "IPv4") return NET_IPV4;
if (net == "ipv6" || net == "IPv6") return NET_IPV6;
```
This alternative implementation would however change the external behaviour of `ParseNetwork(std::string)`.

This pull requests includes a unit test to validate the implementation of `ParseNetwork(std::string)`  prior and after the removal of the `case_conv.hpp` dependency.

`boost/algorithm/string/case_conv.hpp` has been removed from the `EXPECTED_BOOST_INCLUDES` in `test/lint/lint-includes.sh` because it is no longer required.

l2a5b1 committed:
Implements ParseNetwork unit test.

This commit implements a unit test that validates the `ParseNetwork(std::string)` implementation in `netbase.cpp`.

practicalswift commented:
Concept ACK modulo …

* Nice to see the `EXPECTED_BOOST_INCLUDES ` array in `test/lint/lint-includes.sh` get shorter
* This is a good opportunity to also shorten the `KNOWN_VIOLATIONS` array in `test/lint/lint-locale-dependence.sh` – please replace the use of locale dependent functions `std::lower` and `std::upper` with locale independent alternatives

Empact commented on src/rpc/server.cpp:
```diff
@@ -192,9 +191,8 @@ std::string CRPCTable::help(const std::string& strCommand, const JSONRPCRequest&
-                    std::string firstLetter = category.substr(0,1);
-                    boost::to_upper(firstLetter);
-                    strRet += "== " + firstLetter + category.substr(1) + " ==\n";
+                    unsigned char firstLetter = std::toupper(static_cast<unsigned char>(category[0]));
```
`std::toupper(category.front())` should work

Empact commented on src/netbase.cpp:
```diff
@@ -38,7 +37,7 @@ static const int SOCKS5_RECV_TIMEOUT = 20 * 1000;
 
 enum Network ParseNetwork(std::string net) {
-    boost::to_lower(net);
+    std::transform(net.begin(), net.end(), net.begin(), [](unsigned char c) -> unsigned char {return std::tolower(c);});
```
You can pass `tolower` directly like so: `std::transform(net.begin(), net.end(), net.begin(), tolower);`

The same doesn't work with `std::tolower` fyi.

Empact replied:
You could cast like so, I believe that has the same effect: `std::transform(net.begin(), net.end(), net.begin(), (int (*)(int)) std::tolower);`

Empact reviewed (commented):
Good to include `<algorithm>`/`<ctype.h>` where appropriate.

Empact commented:
Concept ACK

laanwj commented:
- is std::tolower locale-independent? (I think so, as @practicalswift mentions, so let's use this opportunity to write a locale-independent deterministic one)
- I'd prefer to factor this out to our own `strtolower` function in `utilstrencodings.cpp`, instead of writing this out here

251Labs commented:
Many thanks for the review comments!

@laanwj 
> is std::tolower locale-independent?

As used in this project at the moment both `std::tolower` and `std::toupper` should be locale-independent and fully deterministic given that the arguments are `std::string` objects with characters in the standard 7-bit ASCII range.

- In one instance `std::toupper` is used to uppercase the first character of `CRPCCommand::category`, which gets its value from arrays declared like the one below where `category` is a `std::string` that uses standard 7-bit ASCII characters:

https://github.com/bitcoin/bitcoin/blob/171b03de065dae94ed6beb5db9ba7f08e92e7c7c/src/rpc/blockchain.cpp#L1923-L1929

- In the second instance it lowercases network names where the accepted values are limited to ["ipv4", "ipv6", "tor", "onion"] to catch a potential  "IPv4" and "IPv6" input.

In this particular case I have also raised the suggestion to drop `boost::to_lower(net)` and replace it with:

```c++
if (net == "ipv4" || net == "IPv4") return NET_IPV4;
if (net == "ipv6" || net == "IPv6") return NET_IPV6;
```

I would even prefer this alternative if the modified external behaviour is acceptable.

@laanwj 
> I'd prefer to factor this out to our own strtolower function in utilstrencodings.cpp, instead of writing this out here

I had definitely considered this but felt that it would be a potential rabbit hole without guidance.

Implementing utility functions `strtolower` and `strtoupper` without the use of the `std::tolower` or `std::toupper` should be trivial if it is acceptable to limit their use to `std::string` objects with standard 7-bit ASCII characters. 

Is this good enough for our purposes?

If the objective is to write custom functions that accept `std::string` objects with extended 8-bit ASCII characters, but can't use `std::tolower` or `std::toupper`, then I am not sure this is doable unless we simplify things tremendously. If the most important thing is to be deterministic, one thing that come to mind is doing conversions based on a single character encoding like ISO-8859-15 and use conversion tables that literally transform characters without considering different alphabets and language specific cases. 

Either way, I am happy to give this another shot by addressing @Empact 's comments; or pursuing alternatives proposed by @practicalswift and @laanwj if I can get further guidance in which direction to take this.

practicalswift commented:
@251Labs For the custom locale independent `std::toupper` functions you would simply return any character outside of the ASCII range as is:

> In the default "C" locale, the following lowercase letters abcdefghijklmnopqrstuvwxyz are replaced with respective uppercase letters ABCDEFGHIJKLMNOPQRSTUVWXYZ.

More generally: as long as your custom functions return the same value as the standard library equivalents would have done under the default "C" locale everything is fine.

251Labs commented:
Thanks again for the feedback!

Per the review comments, I have implemented custom `ToLower` and `ToUpper` functions in 6b2d814 that are locale independent and ASCII deterministic.

I have tried to make the `ToLower(unsigned char)` and `ToUpper(unsigned char)` functions defined in `utilstrencodings.h` as efficient as possible. I hope I did not miss something (obvious) that makes it a failed attempt.

sipa commented on src/test/util_tests.cpp:
```diff
@@ -1202,4 +1202,63 @@ BOOST_AUTO_TEST_CASE(test_DirIsWritable)
 
+BOOST_AUTO_TEST_CASE(test_ToLower)
+{
+    BOOST_CHECK(ToLower('@') == '@');
```
Use `BOOST_CHECK_EQUAL(ToLower('@'), '@')` (it gives a clearer diagnostic message in case of failure).

251Labs replied:
Thanks, addressed in 0234af5

sipa commented on src/utilstrencodings.h:
```diff
@@ -173,4 +173,91 @@ bool ConvertBits(const O& outfn, I it, I end) {
+ */
+inline unsigned char ToLower(unsigned char c)
+{
+    static const unsigned char mappings[256] = {
```
That's a lot of code!

What about `return (c >= 'A' && c <= 'Z' ? (c - 'A') + 'a' : c);`?

251Labs replied:
Thanks @sipa, 

I asked myself the same question and I am happy to change it!

Just to share why I decided to use a mappings array:

**Support for extended ASCII**

Should there be a time where it becomes important to have support for characters in the extended ASCII range then it will be trivial to implement it. It will then *just* be a matter of adjusting entries in mappings array vs extending the implementation with control statements like `if (c == 'Ü') return 'ü';` or alternative logic.

**Performance**

I wanted our locale independent functions to be a good citizen in terms of performance for code that requires it.

I did some naive tests to get an impression what the code would compile to and it seemed that the compiler was able to generate more efficient code in the case `return mappings[c];`

At all optimization levels I would get a pointer to the mappings array where mappings are performed on:
```
lea rax, [rip + __ZZ7ToLowerhE15tolowermappings]
movzx   eax, byte ptr [rdi + rax]
```
Alternatively in the case of `(c >= 'A' && c <= 'Z' ? (c - 'A') + 'a' : c)` the compiler would generate something like this at `O3` optimization:
```
    mov eax, edi 
    add al, -65 
    cmp al, 26
    jae LBB3_2
    add dil, 32
LBB3_2:
    movzx   eax, dil
```
Particularly in the case of iterations I felt the compiler was able to do better optimizations,  for example minimizing iterations by operating in sequences. In the case of the former it could do 4 mapping operations per sequence; while in the latter of at most two mappings per iteration.

On this basis of this I assumed the code would also perform better, but I realize I can be very wrong because I am no expert at this.

I hope this feedback is useful. It it isn't, I will gladly change the code!

251Labs replied:
Thanks @practicalswift, will update the code per @sipa's feedback!

Empact replied:
I agree with @sipa. If we need to add extended ascii support, we can address that then, and this is not a performance-critical operation that calls for extra code to optimize it. Simple, repetitive code adds visual noise and can hide bugs as well.

251Labs replied:
Thanks @empact, I will update the pull request!

251Labs replied:
Thanks again for the feedback. I have updated the pull request and addressed your comments in 53b3bfa.

sipa commented on src/utilstrencodings.h:
```diff
@@ -173,4 +173,91 @@ bool ConvertBits(const O& outfn, I it, I end) {
+ * characters in the standard 7-bit ASCII range.
+ * @param[in,out] str   the string to convert to uppercase.
+ */
+void ToUpper(std::string& str);
```
I feel it's slightly confusing to have a version of `ToUpper` that operates on chars (and does not modify the argument) while the version operating on vectors does modify. Perhaps call this `ConvertToUpper` or so?

251Labs replied:
Thanks, addressed in 0234af5

Empact commented on src/utilstrencodings.h:
```diff
@@ -173,4 +173,91 @@ bool ConvertBits(const O& outfn, I it, I end) {
+ * @return          the uppercase equivalent of c; or the argument
+ *                  if no conversion is possible.
+ */
+inline unsigned char ToUpper(unsigned char c)
```
nit: `constexpr`

Empact commented on src/utilstrencodings.h:
```diff
@@ -173,4 +173,91 @@ bool ConvertBits(const O& outfn, I it, I end) {
+ * @return          the lowercase equivalent of c; or the argument
+ *                  if no conversion is possible.
+ */
+inline unsigned char ToLower(unsigned char c)
```
nit: `constexpr`

Empact commented on src/utilstrencodings.cpp:
```diff
@@ -544,3 +545,18 @@ bool ParseFixedPoint(const std::string &val, int decimals, int64_t *amount_out)
+    std::transform(str.begin(), str.end(), str.begin(), [](unsigned char c){return ToLower(c);});
+}
+
+void ConvertToUpper(std::string& str)
```
I don't think we should define `ConvertToUpper` if we're not using it. The construction can be trivially re-created from `ConvertToLower` if we need it, and in the meantime it's just dead code.

practicalswift commented on src/test/util_tests.cpp:
```diff
@@ -1202,4 +1202,33 @@ BOOST_AUTO_TEST_CASE(test_DirIsWritable)
+    BOOST_CHECK_EQUAL(ToLower('Z'), 'z');
+    BOOST_CHECK_EQUAL(ToLower('['), '[');
+    BOOST_CHECK_EQUAL(ToLower(0), 0);
+    BOOST_CHECK_EQUAL(ToLower(255), 255);
```
Please add some `ToUpper(...)` tests as well :-)

Empact commented on src/rpc/server.cpp:
```diff
@@ -192,9 +191,7 @@ std::string CRPCTable::help(const std::string& strCommand, const JSONRPCRequest&
-                    std::string firstLetter = category.substr(0,1);
-                    boost::to_upper(firstLetter);
-                    strRet += "== " + firstLetter + category.substr(1) + " ==\n";
+                    strRet += "== " + Capitalize(category) + " ==\n";
```
Since this modifies the `category` in-place, I think this breaks the comparison logic, e.g. on line 189.
Sorry, I think my original nit was misguided.

practicalswift replied:
`Capitalize(…)` (`std::string Capitalize(std::string str)`) does not modify `category` in-place :-)

251Labs replied:
Hi @Empact, we should be good. When I addressed your feedback, I changed the method signature from `void Capitalize(std::string&)` (prior to dff017f) to `std::string Capitalize(std::string)` which passes the argument by value (`category` is copied into `str`).

Empact replied:
Thanks, I misread - agree the argument is passed by value, so it's fine. 👍

Empact commented:
utACK 76d70e5

l2a5b1 committed:
Removes the Boost case_conv.hpp dependency.

This commit removes the `boost/algorithm/string/case_conv.hpp` dependency from the project. It replaces the `boost::to_lower` and `boost::to_upper` functions with custom functions that are locale independent and ASCII deterministic.

l2a5b1 committed:
Implements custom tolower and toupper functions.

This commit implements custom equivalents for the C and C++ `tolower` and `toupper` Standard Library functions.
In addition it implements a utility function to capitalize the first letter of a string.

251Labs commented:
b193d5a has been rebased on master (aa39ca764578a9017e03796b554955796022eb0d), woot!

laanwj commented on src/test/netbase_tests.cpp:
```diff
@@ -302,4 +302,22 @@ BOOST_AUTO_TEST_CASE(netbase_getgroup)
+    BOOST_CHECK_EQUAL(ParseNetwork("TOR"), NET_ONION);
+
+    BOOST_CHECK_EQUAL(ParseNetwork(":)"), NET_UNROUTABLE);
+    BOOST_CHECK_EQUAL(ParseNetwork("tÖr"), NET_UNROUTABLE);
```
lol, tÖr, that's like the Swedish version of Tor right?

251Labs replied:
Lol @laanwj, best review comment ever! I do hidden messages or easter eggs from time to time, but this isn't one of them 😂😂

laanwj commented:
utACK b193d5a443bfd994936ad21b807b2bb37756ef2c 
certainly an improvement
